adserver/forms.py

```diff
@@ -1150,6 +1157,190 @@ class Meta:
+                raise forms.ValidationError(
+                    _("Could not retrieve image '%(image)s'."),
+                    params={"image": image_url},
+                )
+
+            image = BytesIO(image_resp.raw.read())
+            width, height = get_image_dimensions(image)
+
+            if width is None or height is None:
+                raise forms.ValidationError(
+                    _("Image for %(name)s isn't a valid image"),
+                    params={
+                        "name": name,
+                    },
+                )
+
+            text_length = len(f"{headline}{content}{cta}")
+
+            for ad_type in self.flight.campaign.allowed_ad_types(
+                exclude_deprecated=True
+            ):
+                log.info(
+                    "Ad-type=%s, required-width=%s, required-height=%s",
+                    ad_type,
+                    ad_type.image_width,
+                    ad_type.image_height,
+                )
+                if ad_type.max_text_length and text_length > ad_type.max_text_length:
+                    raise forms.ValidationError(
+                        _(
+                            "Total text for '%(ad)s' must be %(max_chars)s or less (it is %(text_len)s)"
+                        ),
+                        params={
+                            "ad": name,
+                            "max_chars": ad_type.max_text_length,
+                            "text_len": text_length,
+                        },
+                    )
+
+                if all(
+                    (
+                        ad_type.has_image,
+                        ad_type.image_width,
+                        ad_type.image_height,
+                        (
+                            width != ad_type.image_width
+                            or height != ad_type.image_height
+                        ),
+                        (
+                            width // 2 != ad_type.image_width
+                            or height // 2 != ad_type.image_height
+                        ),
+                    )
+                ):
+                    raise forms.ValidationError(
+                        _(
+                            "Images must be %(required_width)s * %(required_height)s "
+                            "(for %(name)s it is %(width)s * %(height)s)"
+                        ),
+                        params={
+                            "name": name,
+                            "required_width": ad_type.image_width,
+                            "required_height": ad_type.image_height,
+                            "width": width,
+                            "height": height,
+                        },
+                    )
+
+            image_name = image_url[image_url.rfind("/") + 1 :]
+            image_path = f"images/{timezone.now():%Y}/{timezone.now():%m}/{image_name}"
+            default_storage.save(image_path, image)
```

Comment:
Do we want to copy the image to storage here, or after the user as saved it? I guess we want to validate we can access the file, and check sizes, so we need to download it anyway. Having a few ads saved that don't get used seems fine.